cordformation/src/main/kotlin/net/corda/plugins/Dockerform.kt

```diff
@@ -146,33 +144,70 @@ open class Dockerform @Inject constructor(objects: ObjectFactory) : Baseform(obj
 
                 it.installDefaultDatabaseConfig(dbConfig)
 
-                val args = mutableMapOf<String, Any>()
-
-                dockerfileArgs.entrySet().toList().map {
-                    args[it.key] = it.value.unwrapped()
-                }
-
-                services[dbHost] = mapOf(
+                val dbDockerfile = dockerConfig.getString("dockerConfig.dbDockerfile")
+                // Override the port and hostname parameters in dockerfile
+                val dbDockerfileArgs = defaultUrlArgs.withFallback(
+                        dockerConfig.getConfig("dockerConfig.dbDockerfileArgs")).resolve()
+                val database = mutableMapOf(
                         "build" to mapOf(
                              "context" to project.buildDir.absolutePath,
                              "dockerfile" to project.buildDir.resolve(dbDockerfile).toString(),
-                             "args" to args
+                             "args" to dbDockerfileArgs.entrySet().map { it.key to it.value.unwrapped() }.toMap()
                         ),
-                        "restart" to "unless-stopped",
-                        "ports" to listOf(dbPort)
+                        "restart" to "unless-stopped"
                 )
 
+                // append persistence volume if it is required
+                if (dockerConfig.hasPath("dockerConfig.dbDataVolume")) {
+                    var hostPath = dockerConfig.getString("dockerConfig.dbDataVolume.hostPath")
+                    var containerPath = dockerConfig.getString("dockerConfig.dbDataVolume.containerPath")
+
+                    if (dockerConfig.hasPath("dockerConfig.dbDataVolume.pathArgs")) {
+                        val pathArgs = dockerConfig.getConfig("dockerConfig.dbDataVolume.pathArgs")
+
+                        hostPath = ConfigFactory.parseString("HOSTPATH=${TypesafeUtils.encodeString(hostPath)}")
+                                .resolveWith(ConfigFactory.empty().withValue("NODENAME", ConfigValueFactory.fromAnyRef(it.nodeDir.name)))
+                                .getString("HOSTPATH")
+                        containerPath = ConfigFactory.parseString("CONTAINERPATH=${TypesafeUtils.encodeString(containerPath)}")
+                                .resolveWith(pathArgs)
+                                .getString("CONTAINERPATH")
+                    }
+
+                    var absoluteHostPath = directoryPath.resolve(hostPath)
+                    val hostDir = File(absoluteHostPath.toUri())
+                    if (hostDir.mkdirs() || hostDir.isDirectory) {
+                        volumes["$dbHost-volume"] = mapOf(
+                                "driver" to "local",
+                                "driver_opts" to mapOf(
+                                        "type" to "none",
+                                        "device" to "$absoluteHostPath",
```

Comment:
I think `"$absoluteHostPath"` is equivalent to `absoluteHostPath.toString()` because you're not asking Kotlin to join any strings together.